zict/lru.py

```diff
@@ -63,59 +71,79 @@ def __init__(
     ):
         self.d = d
         self.n = n
-        self.order = dict.fromkeys(d)
         if callable(on_evict):
             on_evict = [on_evict]
         self.on_evict = on_evict or []
         self.weight = weight
         self.weights = {k: weight(k, v) for k, v in d.items()}
-        self.total_weight = sum(self.weights.values())
-        while self.total_weight > n:
-            self.evict()
+        self._total_weight = Accumulator(sum(self.weights.values()))
+        self.order = InsertionSortedSet(d)
+        self.heavy = InsertionSortedSet(k for k, v in self.weights.items() if v >= n)
+        self.closed = False
+
+    @property
+    def total_weight(self) -> float:
+        return self._total_weight.value
 
     def __getitem__(self, key: KT) -> VT:
         result = self.d[key]
-        try:
-            del self.order[key]
-        except KeyError:
-            # Race condition which can happen during multithreaded access
-            pass  # pragma: nocover
-        self.order[key] = None
+        # Don't use .remove() to prevent race condition which can happen during
+        # multithreaded access
+        self.order.discard(key)
+        self.order.add(key)
         return result
 
     def __setitem__(self, key: KT, value: VT) -> None:
+        self.set_noevict(key, value)
+        try:
+            self.evict_until_below_capacity()
+        except Exception:
+            if self.weights[key] > self.n and key not in self.heavy:
+                # weight(value) > n and evicting the key we just inserted failed.
+                # Evict the rest of the LRU instead.
+                try:
+                    while len(self.d) > 1:
+                        self.evict()
+                except Exception:
+                    pass
```

Comment:
The behaviour implemented by blurb is very opinable. It's there mostly to retain functional compatibility with main (there are several tests in dask/distributed that rely on this behaviour too).
Ultimately, once we transition to async set, this will become dead code (at least in distributed).